feat(ivy): runtime i18n

ocombe opened the pull request:
## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Feature
```

## What is the current behavior?
There is no runtime i18n for ivy


## What is the new behavior?
Runtime i18n works for ivy
This is only the first step: html and expressions.
We still need to do ICUs, projected content, and the service for code translations.


## Status
Done:
- translate HTML elements (including moving/removing elements)
- translate expressions (including moving/removing expressions)
- translate attributes (including with expressions)
- support multiple i18n blocks
- support `container` instructions (with or without embedded template, including moving/removing them)
- support `ng-container`
- support projections (including with selectors, and multiple projections)
- canonical tests for html/expressions/expressions in attributes for the compiler

To do:
- updating design spec with recent changes


## Does this PR introduce a breaking change?
```
[x] No
```

mary-poppins commented:
You can preview 391ac36 at https://pr24037-391ac36.ngbuilds.io/.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,116 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+  MASK = ~((1 << 29) - 1),
+}
+
+export function i18nMapping(msg: string, placeholders: string[]): any[] {
```
can you type the return value better than `any[]` maybe `I18nInstruction[]` where `type Instruction = number|string`

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,116 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+  return instructions;
+}
+
+export function i18nApply(instructions: any[]): void {
```
`any[]` => `I18nInstruction[]`

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,116 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+  // TODO(ocombe): find a way to cache this
+  const children = getChildren(previousOrParentNode);
+  for (let i = 0; i < children.length; i++) {
+    children.splice(i + 1, 0, ...getChildren(children[i]));
```
can you not use `...` operators since when down-leveled to `ES5` it is slow.
Notice that it causes `apply` to be called
http://www.typescriptlang.org/play/#src=children.splice(i%20%2B%201%2C%200%2C%20...getChildren(children%5Bi%5D))%3B

ocombe replied:
I can use slice & concat instead: https://jsperf.com/slice-concat-vs-splice-spread
It's faster in chrome (about the same in firefox).
Unless you have a better idea?

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,116 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+  if (args.some(value => bind(value) !== NO_CHANGE)) {
+    const expRegex = /\{\$(exp_\d+)\}/g;
+    return msg.replace(
+        expRegex, (match: string, p1: string) => { return args[placeholders.indexOf(p1)]; });
```
You are allocating a regexp and than doing string replace. This will create GC pressure every time we run change detection and it is not the fastest. Can we refactor this so that it does not allocate anything when there is no change to inputs.

ocombe replied:
actually this is the same regex that it used above, I'll use the same constant

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,116 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+  }
+}
+
+export function i18nInterpolation(
```
What is the point of this method with 7 args? Why 7?

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,116 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+/** A variable length version of i18nInterpolation */
+export function i18nInterpolationV(msg: string, placeholders: string[], args: any[]): string|
+    NO_CHANGE {
+  if (args.some(value => bind(value) !== NO_CHANGE)) {
```
Don't use `some` same argument as `forEach`

mhevery commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -456,6 +456,28 @@ export function appendChild(parent: LNode, child: RNode | null, currentView: LVi
   return false;
 }
 
+export function getChildren(parent: LNode): LNode[] {
```
The way you can make this method cacheable, is to collect method induces rather than `LNode`s themselves. So return would be `number[]`. Than you could do the dereference when processing instructions.

mary-poppins commented:
You can preview 767144e at https://pr24037-767144e.ngbuilds.io/.

mary-poppins commented:
You can preview b4a00f4 at https://pr24037-b4a00f4.ngbuilds.io/.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,267 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+    const instruction = instructions[i] as number;
+    switch (instruction & I18nFlags.INSTRUCTION_MASK) {
+      case I18nFlags.ELEMENT:
+        const element = children[instruction & I18nFlags.INDEX_MASK];
```
`data[children[instruction & I18nFlags.INDEX_MASK]]`

`data[instruction & I18nFlags.INDEX_MASK]`

mhevery commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -456,6 +456,28 @@ export function appendChild(parent: LNode, child: RNode | null, currentView: LVi
+  const children = [];
+  let currentNode = parent.child;
+  while (currentNode) {
+    if (currentNode.native) {
```
`currentNode.tNode.index`

JoostK commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,267 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+ */
+export function i18nInterpolation2(msg: string, ph0: string, ph1: string, a0: any, a1: any): string|
+    NO_CHANGE {
+  if (bind(a0) !== NO_CHANGE || bind(a1) !== NO_CHANGE) {
```
We could use `bindingUpdatedN` here, with `N` in `{,2,4}` to combine the bind checks. That would also get rid of the need for `NO_CHANGE` as then only booleans are handled. Refer to the `interpolation#` instructions where the same approach is used.

Applies to all `i18nInterpolation#` functions.

JoostK replied:
Also, watch out with short circuiting. We do not want that to happen, because it would cause bindings not to be updated. Recall that `bind` wither updates or consumes a binding which causes `bindingIndex` to be incremented past the binding slot, we need to make sure that always occurs for the bindings to stay in sync.

ocombe replied:
Good points, I didn't think about that! thanks

ocombe replied:
actually I think I should probably use interpolation1, ... instead

JoostK commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,267 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+    NO_CHANGE {
+  for (let i = 0; i < args.length; i++) {
+    if (bind(args[i]) !== NO_CHANGE) {
+      return msg.replace(
```
The early return here can't be right. We'd need to traverse all bindings for them to be updated and become string replaced.

JoostK commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,267 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+  for (let i = 0; i < msgList.length; i++) {
+    const value = msgList[i];
+    let match;
+    if (match = elementRegex.exec(value)) {
```
We can probably get away with a lot less regex matching here, if we were to change the regex into `/\{\$((?:exp|p|c)_\w+)\}/g`. We'd then know that all odd indices actually match, and we can easily figure out the instruction per the odd matches' prefix and the identifier from a substring. One regex match, then just string manipulation with known offsets.

ocombe replied:
Very good idea, thanks!

JoostK commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2436,38 +2443,26 @@ export function i18nMapping(msg: string, placeholders: string[]): I18nInstructio
-    childrenIndex = previousOrParentNode.tNode !.childrenIndex !;
-  }
-
+  let children = data.slice(startIndex, endIndex + 1);
```
If we offset each lookup into `data` with `startIndex`, we'd void having to create a copy. We may then even get rid of `endIndex`, or use it only in dev mode for bounds checking.

mary-poppins commented:
You can preview 1171f4c at https://pr24037-1171f4c.ngbuilds.io/.

mary-poppins commented:
You can preview b528710 at https://pr24037-b528710.ngbuilds.io/.

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+
+/**
+ * A list of flags to encode the i18n instructions used to translate the template.
+ * We shift the flags by 29 so that 30 & 31 & 32 bits contains the instructions.
```
nit: use constant rather than number in the comment and in the enum to avoid then getting out of sync later

ocombe replied:
I tried that at first but for some reason typescript doesn't want a variable in a an enum value

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+  /** Used to decode the number encoded with the instruction. */
+  INDEX_MASK = (1 << 29) - 1,
+  /** Used to test the type of instruction. */
+  INSTRUCTION_MASK = ~((1 << 29) - 1),
```
nit: `0b111 << INSTRUCTION_SHIFT`

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+ * and returns a list of instructions that will be used to translate the template.
+ */
+export function i18nMapping(msg: string, placeholders: string[]): I18nInstruction[] {
+  const instructions = [];
```
add type ?

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+
+  for (let i = 0; i < msgList.length; i++) {
+    const value = msgList[i];
+    // odd index is either element, expression or close tag
```
`// odd index are placeholders`

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+    const value = msgList[i];
+    // odd index is either element, expression or close tag
+    if ((i % 2) === 1) {
+      switch (value.substr(0, 1)) {
```
`value.charAt(0)`

_Note: p/e/c will change later_

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+    if ((i % 2) === 1) {
+      switch (value.substr(0, 1)) {
+        case 'p':
+          instructions.push(placeholders.indexOf(value) | I18nFlags.ELEMENT);
```
why use `.indexOf(value)` instead of the map as proposed in the compiler side spec ?

ocombe replied:
good point, I've changed the code a few times, and now it seems like a map would be better, I'll update my code

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+          break;
+      }
+    } else {  // it's a string
+      // don't create a text node for empty values
```
I don't understand what `// don't create a text node for empty values` means here ?
Should it be a TODO ? / What not doing it then ?

ocombe replied:
When using split above, it will create empty string fragments, for example: `{$p_1}trad 1{$c_p}` will be `["", "p_1", "trad 1", "c_p", ""]. The `if` here is used to avoid pushing instructions that are useless, we don't want to create empty text node

vicb replied:
Oh you are absolutely right. I totally miss that there was an if here.

May be write 
``` typescript
} else if (value) {
  instructions.push(I18nFlags.TEXT, value);
}
```

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+  for (let i = 0; i < placeholders.length; i++) {
+    const ph = placeholders[i];
+    if (phAdded.indexOf(ph) === -1) {
+      instructions.push(i | I18nFlags.REMOVE_NODE);
```
You should not push instructions for closing tags ?

ocombe replied:
when we read the instructions, we need to know the current parent to which we should add the childs, we change the parent when we see an opening tag (`p_x`), and change it again when we see the closing tag

vicb replied:
Yes I agree. But this seems to be `CLOSE_NODE` l2433, this one is `REMOVE_NODE`

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+
+/** Takes a list of instructions generated by `i18nMapping` to transform the template accordingly */
+export function i18nApply(startIndex: number, instructions: I18nInstruction[]): void {
+  ngDevMode &&
```
nit:
I find
```
if (ngDevMode) {
  // 2 instrcutions
}
```
more readable ?

ocombe replied:
it's like that all over the place in `instructions.ts`, I don't mind changing it, since it's one less test to do, but we should change it in the other parts of the code as well

vicb replied:
Different people like to write code in different ways. I find the proposal above easier to read.
When I say "nit:" it's a hint that you might or might not want to do it. Up to what you find is more readable.

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+  ngDevMode && ngDevMode.rendererMoveNodes++;
+
+  let localPreviousOrParentNode = previousOrParentNode;
+  for (let i = 0; i < instructions.length; i++) {
```
As discussed with @mhevery, not moving the LNode will break projections.
Has this been thought through in the design spec ?

ocombe replied:
no, that's why I asked if we could have a meeting about that, but @mhevery said that he would think about it, I'm giving him some time for that first, I'll update the code once we have a good solution

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+    const instruction = instructions[i] as number;
+    switch (instruction & I18nFlags.INSTRUCTION_MASK) {
+      case I18nFlags.ELEMENT:
+        const element = data[startIndex + instruction & I18nFlags.INDEX_MASK];
```
Question:
why `startIndex + instruction & I18nFlags.INDEX_MASK` ?
In my compiler side spec the index (` instruction & I18nFlags.INDEX_MASK`) is an absolute index.

ocombe replied:
ok, I'll change my code then

vicb replied:
Also because the index only has a limited number of bits for encoding, there should be an assert to make sure that no overflow occurs

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+ * @param ph0 the name of the expression in the translation
+ * @param v0 the value of the expression
+ */
+export function i18nInterpolation1(msg: string, ph0: string, v0: any): string|NO_CHANGE {
```
The compiler design spec propose and implementation with no need for the 1, 2, ..., 8 variants.
Why do you think this one is better ?

vicb replied:
Oh this is used for attributes only.

Need to be discussed.

ocombe replied:
isn't it more optimized to do it like that?

vicb replied:
My spec is for the compiler side:

> This design specs details the output of the compiler for i18n templates. It does also describe what the IVY runtime should do but not how it should do it.

vicb replied:
What I envision:
```
function i18nInterpolation(
   mapping: I18nMapping, 
   a0: any, 
   a1: any, 
   // ... 
   a7: any
): string | NO_CHANGE;
```

The first argument `mapping` could contain a list of static strings `['a', 'b', 'c']` and a list of placeholder index `[2, 0, 1]` compute once the first time the template function is executed.

Updating the value would compute `'a' + a2 + 'b' + a0 + 'c' + a1`.

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+ * A little bit more efficient than using replace and instantiating a new RegExp.
+ * See: https://jsperf.com/replace-regex-split-join.
+ */
+function replaceAllStr(str: string, pattern: string, replacement: string): string {
```
I am not sure to understand how `replaceAllStr()` / ` i18nInterpolationX()` will work at all.
Seems like this is unused in this PR / not described in the doc.
Remove from this PR for now until details are figured out ?

vicb reviewed (dismissed):
Great start.

ocombe commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+
+/** Represents the instructions used to translate the template. */
+export type I18nInstruction = number | string;
+const tagRegex = /\{\$((?:exp|p|c)_\w+)\}/g;
```
I don't see how to split the translation efficiently without a regex?

vicb replied:
There should probably be a regexp, but a simpler/faster one ie `/\{\$[^}]+\}/g`
The other part of the perf gain is that placeholders would be specified as maps (one for element and an other one for expression).

ocombe commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2374,3 +2372,302 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+  for (let i = 0; i < msgList.length; i++) {
+    const value = msgList[i];
+    // odd index is either element, expression or close tag
+    if ((i % 2) === 1) {
```
oh nice trick

mary-poppins commented:
You can preview 2d23955 at https://pr24037-2d23955.ngbuilds.io/.

mary-poppins commented:
You can preview 8203e35 at https://pr24037-8203e35.ngbuilds.io/.

mary-poppins commented:
You can preview 81c65ad at https://pr24037-81c65ad.ngbuilds.io/.

mary-poppins commented:
You can preview 95d5e5c at https://pr24037-95d5e5c.ngbuilds.io/.

JoostK commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2572,187 +2572,90 @@ export function i18nApply(
+  const staticText: I18nInstruction[] = translation.split(tagRegex);
+  for (let i = 0; i < staticText.length; i++) {
+    // odd indexes are placeholders
+    if (i & 1) {
```
`i += 2` would avoid the odd check.

mary-poppins commented:
You can preview c0b6c98 at https://pr24037-c0b6c98.ngbuilds.io/.

mary-poppins commented:
You can preview 3da714a at https://pr24037-3da714a.ngbuilds.io/.

ocombe commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -77,6 +76,10 @@ export function getNextLNode(node: LNode): LNode|null {
 
 /** Retrieves the first child of a given node */
 export function getChildLNode(node: LNode): LNode|null {
+  // for i18n, we use pNextOrParent to point to the virtual next child
```
I decided to use `pNextOrParent` to point to the first child for projections, but I could also add another new property on the `LNode` or `TNode` (like `pChild` for example) that would be used just for that. What do you prefer @mhevery @vicb?

vicb replied:
I don't know this code well enough to tell.
What is a "virtual" child ?

mhevery replied:
What you did is correct.

However, @kara is refactoring to get rid of `LNode`s and so your approach will have to change. @kara do you think we can leave it now, and deal with it when we do our refactoring or should we design it differently now?

The new approach I think will have to create a `T` array for the projected nodes instead of `pNextOrParent`, which means that these new nodes will have to be stored someplace in the data array. Seems like you should just go with this approach, and we will have to fix it once we do the refactoring.

kara replied:
Yeah, this does create a bit more work for me, but I don't think it makes sense to do it now with the data refactoring 50% done. Can we put a TODO in here to remove?

ocombe commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1782,7 +1781,7 @@ export function projectionDef(
     }
 
-    componentChild = getNextLNode(componentChild);
+    componentChild = getNextLNodeWithProjection(componentChild);
```
I haven't found any downside to use `getNextLNodeWithProjection` instead of `getNextLNode`, and all the tests pass, but maybe there something that I missed here?

pkozlowski-opensource replied:
@vicb you are overestimating my knowledge here :-) 

Anyhow, I find this change slightly surprising since the oryginal logic was:
* go over each and every content node;
* distribute it (potentially adding it to the "default" bucket if none of the selector matches)

With this change it seems like we can skip / omit some of the nodes when going over content nodes and on the surface it doesn't sound right. What is the motivation for this change?

ocombe commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2452,3 +2451,234 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+    }
+  }
+
+  // TODO(ocombe): add a test to check if we really need to update the last pointer
```
I'm not sure if we really need to update the last node pointer, since it should be the parent (the i18n block), and it shouldn't change anything for projections. If that's the case, then we can drop `endIndex` because it isn't used anywhere.
Can you come up with a scenario where updating that pointer is necessary?

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2452,3 +2451,234 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+    if (i & 1) {
+      switch (value.charAt(0)) {
+        case 'p':
+          instructions.push(placeholders[value] | I18nFlags.Element);
```
Add assert to make sure `placeholders[value]` does not overflow

ocombe replied:
I think that we can safely assume that we will never have an index number that exceeds 28 bits (268435456), right?

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2452,3 +2451,234 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+/**
+ * Takes a list of instructions generated by `i18nMapping` to transform the template accordingly
+ *
+ * @param startIndex Used for content projections
```
May be start by what this is before how it is used ?

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2452,3 +2451,234 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+        ngDevMode && ngDevMode.rendererCreateTextNode++;
+        const value = instructions[++i];
+        const textRNode = createTextNode(value, renderer);
+        // create the LNode for projections
```
Explain why we need this ?

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2452,3 +2451,234 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+}
+
+/** Represents the instructions used to translate the template. */
+export type I18nInstruction = number | string;
```
Expand on what instructions could be:
- a placeholder index,
- a static text,
- a bit field

vicb commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -0,0 +1,327 @@
+          }
+          if (rf & RenderFlags.Update) {
+            textBinding(1, bind(ctx.exp1));
+            textBinding(2, bind(ctx.exp2));
```
- Do you have a test with a `textBinding` to a node that has been remove ?

vicb commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -0,0 +1,327 @@
+    // html tags are replaced by placeholders.
+    // open tag placeholders are never re-used (closing tag placeholders can be).
+    const MSG_1 = `{$p_3}trad 1{$c_p}{$p_0}trad 2{$p_1}trad 3{$c_p}{$c_p}`;
+    const i18n_1 = i18nMapping(MSG_1, {p_0: 1, p_1: 2, p_2: 3, p_3: 4});
```
seems like closing tag placeholders are missing (in multiple tests)

vicb commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -0,0 +1,327 @@
+          if (rf & RenderFlags.Create) {
+            elementStart(0, 'div');
+            // Start of translated section
+            text(1);
```
For each test, you should probably explicit the source structure and the translation.
Also explain what the `text()` instructions are for ? I don't see why we have 3 here ?

vicb commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -0,0 +1,327 @@
+
+  it('should support both html elements, interpolations and interpolations in attributes', () => {
+    const MSG_1 = `{$exp_1} {$p_3}trad {$exp_2}{$c_p}`;
+    const MSG_2 = `start {$exp_2} middle {$exp_1} end`;
```
may be use longer name to give more context ? 
- `MSG_ATTRIBUTE_1`
- `MSG_DIV_2`
- ...

ocombe commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2452,3 +2451,234 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+ * value will be concatenated into the final translation.
+ */
+export function i18nExpMapping(
+    translation: string, placeholders: PlaceholderMap = {}): I18nInstruction[] {
```
we can just say that it cannot be empty, and pass an empty object when there is no placeholder

ocombe commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2452,3 +2451,234 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+ */
+export function i18nExpMapping(
+    translation: string, placeholders: PlaceholderMap = {}): I18nInstruction[] {
+  const staticText: I18nInstruction[] = translation.split(tagRegex);
```
Using more local variables is less efficient, you'd need to loop on all indexes instead of just `odd` ones.
I'll add a new type since it's not the same kind of instructions (no binary values here)

ocombe commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2452,3 +2451,234 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+    let value: any;
+    // odd indexes are placeholders
+    if (i & 1) {
+      switch (instructions[i]) {
```
it's 30% slower with an external function https://jsperf.com/switch-function-inline

mhevery replied:
leave as is than.

ocombe commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2452,3 +2451,234 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
+          break;
+      }
+
+      if (bindingUpdated(value)) {
```
I'll have to check the bindings for v0 to v7, even if we only pass 1 value, since there's no way to know if the value is undefined or if we don't have that value at all...

mary-poppins commented:
You can preview 7185151 at https://pr24037-7185151.ngbuilds.io/.

kara reviewed (commented):
Some nits about comments, if you have time.

mary-poppins commented:
You can preview 4b671b4 at https://pr24037-4b671b4.ngbuilds.io/.

mary-poppins commented:
You can preview a95898b at https://pr24037-a95898b.ngbuilds.io/.

mary-poppins commented:
You can preview 6c566bb at https://pr24037-6c566bb.ngbuilds.io/.

vicb reviewed (changes requested):
- There are deviations from the compiler side spec to be discussed before this can go in,
- Not all my comments have been addressed,
- Please do not dismiss my reviews w/o asking me before.

mhevery reviewed (changes requested):
Waiting until the container implementation is done

mary-poppins commented:
You can preview 1106b3a at https://pr24037-1106b3a.ngbuilds.io/.

mary-poppins commented:
You can preview 82544b7 at https://pr24037-82544b7.ngbuilds.io/.

mary-poppins commented:
You can preview 18f2153 at https://pr24037-18f2153.ngbuilds.io/.

mary-poppins commented:
You can preview 27665ad at https://pr24037-27665ad.ngbuilds.io/.

vicb commented:
- Please add canonical tests in `packages/core/test/render3/compiler_canonical` (ie tests that will get mirrored in the compiler tests),
- Seems like all the projection tests use NO selector. I would expect to see tests trying to project element that have been removed explicitly ?

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,340 @@
+ * A function to recursively generate the instructions for i18nMapping
+ *
+ * @param index The current index in msgList
+ * @param msgList The translation string split into an array of placeholders and text elements
```
not in sync with the code

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,340 @@
+export const tagRegex = /\{\$([^}]+)\}/g;
+
+/**
+ * A function to recursively generate the instructions for i18nMapping
```
> `A function to recursively generate the instructions for i18nMapping`
doesn't really help to know the intent of this function.
What are the inputs, the outputs, what does this fn works on ? a template ?
"to recursively" seems useless here. That's purely an implementation details that brings not useful information to the caller.

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,340 @@
+    if (index & 1) {
+      let phIndex;
+
+      if (elements && elements[tmplIndex] &&
```
Add a comment on what  you are testing and what instruction you generate. (one liner)

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,340 @@
+        tmplInstructions.push(phIndex | I18nFlags.Element);
+        phAdded.push(value);
+
+        if (tmplIndex > 0 || (tmplContainers && tmplContainers.indexOf(value) !== -1)) {
```
where is the spec for the handling of embedded template ? I thought you told me you would update it ?

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,340 @@
+ */
+function generateInstructions(
+    index: number, instructions: I18nInstruction[][], msgList: string[],
+    elements: (PlaceholderMap | null)[] | null, expressions?: (PlaceholderMap | null)[] | null,
```
Question: why and array of `PlaceholderMap` ?

ocombe replied:
The indexes of the placeholders are different in each template function, we need a placeholder map for each template because of that

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  if (elements) {
+    const tmplElements = elements[tmplIndex];
```
comment, comment, comment...
what are you trying to do ?

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,340 @@
+}
+
+/**
+ * Takes a translation string and the initial list of placeholders (elements and expressions)
```
Please separate and clearly indicate in the comment (again)
1) instructions, called from template
2) internal functions

ocombe committed:
feat(ivy): runtime i18n

ocombe committed:
fixup! feat(ivy): runtime i18n

mary-poppins commented:
You can preview 4202fcf at https://pr24037-4202fcf.ngbuilds.io/.

ocombe commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -0,0 +1,1094 @@
+              '<ul><li><span>valeur: 1!</span><span>valeur: 2!</span></li><li><span>valeur: 1!</span><span>valeur: 2!</span></li></ul>');
+
+      // TODO(ocombe): there's a bug with embedded containers, uncomment once it has been fixed
+      // Remove the last item
```
PR to fix that bug: https://github.com/angular/angular/pull/24564

mhevery replied:
I will merge it in a bit, please resolve this.

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+export const tagRegex = /\{\$([^}]+)\}/g;
+
+/**
+ * Internal function that reads the translation parts and generates a set of instructions for each
```
Could at least you move the i18n API first and the internal fn last (if possible different files would be better)

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+ * @param elements An array containing, for each template, the maps of element placeholders and
+ * their indexes.
+ * @param expressions An array containing, for each template, the maps of expression placeholders
+ * and their indexes.
```
Could you please add more detail on element ph indexes / expression ph indexes ? (may be adding a small ex)

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+ * when generating the instructions for their parent template.
+ * @param lastChildIndex The index of the last child of the i18n node. Used when the i18n block is
+ * an ng-container.
+ * @returns the current index in msgList
```
what is `msgList` ?

may be giving an example would help understand this function ?

ocombe replied:
aren't the tests enough?

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+ *
+ * See `i18nMapping()` for more details.
+ *
+ * @param index The current index in `translationParts`.
```
if index applies to `translationParts`, it would be nice to have those 2 back to back ?

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+    if (index & 1) {
+      let phIndex;
+
+      // Test if the value is an element
```
"Test if" sounds useless. Move the comment inside the `if`: "The placeholder represents a DOM element`

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+      // Test if the value is an element
+      if (elements && elements[tmplIndex] &&
+          typeof(phIndex = elements[tmplIndex] ![value]) !== 'undefined') {
+        // Add an instruction to move the element to that index
```
"that index" ?

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+    tmplContainers?: string[] | null, lastChildIndex?: number | null): number {
+  const tmplIndex = instructions.length;
+  const tmplInstructions: I18nInstruction[] = [];
+  const phAdded = [];
```
`phVisited` ?

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+        tmplInstructions.push(phIndex | I18nFlags.Element);
+        phAdded.push(value);
+
+        if (tmplIndex > 0 || (tmplContainers && tmplContainers.indexOf(value) !== -1)) {
```
`tmplIndex > 0`:
- I have no clue why this test,
- because `tmplIndex` is a const, you should move the test oustise of the loop `const is... =  tmplIndex > 0`;

ocombe replied:
Actually the "if" is useless here, we can just always increase the `tagCounter`

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+ * their indexes.
+ * @param expressions An array containing, for each template, the maps of expression placeholders
+ * and their indexes.
+ * @param tmplContainers An array of template container placeholders whose content should be ignored
```
what are `template container` ?

ocombe replied:
I'm talking about instructions of the type `container(1, liTemplate, null, ['ngForOf', '']);` that have a template function and generate an embedded template.

vicb replied:
view containers

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+          expressions && expressions[tmplIndex] &&
+          typeof(phIndex = expressions[tmplIndex] ![value]) !== 'undefined') {
+        // Add an instruction to move the expression to that index
+        tmplInstructions.push(phIndex | I18nFlags.Expression);
```
how can we "move" an expression ?

ocombe replied:
by moving the text node associated with this expression

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+  const tmplIndex = instructions.length;
+  const tmplInstructions: I18nInstruction[] = [];
+  const phAdded = [];
+  let tagCounter = 0;
```
`tagConter` -> `openedTagCount` ?

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+        if (phAdded.indexOf(ph) === -1) {
+          let index = tmplExpressions[ph];
+          // Add an instruction to remove the expression
+          tmplInstructions.push(index | I18nFlags.RemoveNode);
```
Question: should RemoveNode be DetachNode ?
Same comment as few weeks ago: assert that the index does not overflow (`ngDevMode && assert(...)`)

ocombe replied:
I added the assert but I feel like it's useless, we will never have an index that goes beyond 28 bits (~268435455)

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+  }
+
+  if (tmplIndex === 0 && typeof lastChildIndex === 'number') {
+    // The current parent has more children, we need to append them at the end to keep the order
```
I don't understand `we need to append them at the end to keep the order` ?

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+    // The current parent has more children, we need to append them at the end to keep the order
+    for (let i = maxIndex + 1; i <= lastChildIndex; i++) {
+      // We consider those additional placeholders as expressions because we don't care about
+      // their children, all we need to do is to append them at the end
```
append means at the end -> you can remove "at the end"

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+
+/**
+ * Takes a translation string, the initial list of placeholders (elements and expressions) and their
+ * indexes, and returns a list of instructions for each template.
```
same comment as above in my design spec, we use expressions order, not indexes. Pls clarify.

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,347 @@
+  const translationParts = translation.split(tagRegex);
+  const instructions: I18nInstruction[][] = [];
+
+  // Call the recursive function that will update the instructions
```
useless comment ?
but maybe the function should be named `updateInstructions` / `appendInstructions` ?

vicb reviewed (changes requested):
See inline commits for feedback.
Please add to this PR:
- What is done and todo in the PR headline,
- The status of the design spec in the PR headline,
- What is changed by each fixup commits.
So that github (and not slack or any other channel) is the single/up to date source of truth.

I can see some progress by I have no clear idea of what is left to be done.

ocombe committed:
fixup! feat(ivy): runtime i18n

mary-poppins commented:
You can preview e2654f0 at https://pr24037-e2654f0.ngbuilds.io/.

vicb commented on packages/core/src/core_render3_private_export.ts:
```diff
@@ -8,40 +8,52 @@
 
 // clang-format off
 export {
-  defineComponent as ɵdefineComponent,
```
Re-ordering alphabetically is probably a good idea... but not in this PR.
(and such refactor should be done in refactor commits)

ocombe replied:
not my fault, it's `gulp format`! And it's needed to pass the lint tests

ocombe replied:
I think I found the mistake, removing the last `,` at the end of the list. I've added it back and it seems to be ok this time.

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,332 @@
+import {stringify} from './util';
+
+/**
+ * A list of flags to encode the i18n instructions used to translate the template.
```
Technically speaking, those are not Flags (a flag is usually a bit that can be 0 or 1).
Check the core to see how they are called where we use such a mechanism an name this one consistently.

mhevery commented on packages/core/test/render3/compiler_canonical/i18n_spec.ts:
```diff
@@ -0,0 +1,93 @@
+            $r3$.ɵE(0, 'div');
+            $r3$.ɵT(1);
+            $r3$.ɵe();
+            $r3$.ɵiA(1, $i18n_1$[0]);
```
Why is `$i18n_1$` an array where we read `[0]`? Under what conditions would we use the other items in the array?

mhevery replied:
I see when the translation crosses container boundaries.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2580,3 +2591,138 @@ export function _getComponentHostLElementNode<T>(component: T): LElementNode {
 export const CLEAN_PROMISE = _CLEAN_PROMISE;
 export const ROOT_DIRECTIVE_INDICES = _ROOT_DIRECTIVE_INDICES;
+
+///////////////////////////////
```
Could we move this section to `i18n.ts`? This file is already to big.

mhevery commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -0,0 +1,1155 @@
+          '<ul><li><span>valeur: one!</span><span>valeur: two!</span></li></ul>');*/
+    });
+
+    // TODO(ocombe): make this work once PR #24346 has landed
```
The PR has landed. Can we remove the comment and resolve it?

mhevery commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -149,6 +156,11 @@ export interface LContainerNode extends LNode {
+   * If this node is part of an i18n block, it indicates whether this container is part of the DOM
+   * If this node is not part of an i18n block, this field is null.
+   */
+  detached: boolean;
```
- The comment above says it is `null` but it is `boolean`
- The bigger problem is that this data is on `LNode` which means that we would have to store it in `LViewData` after @kara is finished refactoring. This needs to move to the `TNode`. Since all templates will have their nodes translated same way it should be fine there. This needs to be resolved before we can merge it in.

mhevery reviewed (changes requested):
I have given it a thorough review and fixed minor issues here b697333535. 

The big issue which needs to be resolved before this can be merged in is moving `detached` from `LNode` to `TNode`.

Also left few other comments.

mary-poppins commented:
You can preview b697333 at https://pr24037-b697333.ngbuilds.io/.

ocombe commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,332 @@
+          typeof(phIndex = elements[tmplIndex] ![value]) !== 'undefined') {
+        if (ngDevMode) {
+          assertLessThan(
+              phIndex.toString(2).length, 28, `Index ${phIndex} is too big and will overflow`);
```
that's what I said to Victor ¯\_(ツ)_/¯

ocombe committed:
fixup! feat(ivy): runtime i18n

ocombe commented:
new fixup commit:
- moved everything into i18n.ts
- fixed reordering of exports for core_render3_private_export.ts
- improved performance for `i18nInterpolationV` and added a test for it
- moved `detached` property to `tNode`
- other small stuff

mary-poppins commented:
You can preview ffc36db at https://pr24037-ffc36db.ngbuilds.io/.

vicb commented on packages/core/src/core_render3_private_export.ts:
```diff
@@ -87,6 +87,13 @@ export {
+  iI as ɵiI,
+  iIV as ɵIV,
+  iM as ɵiM,
+  I18nInstruction as ɵI18nInstruction,
```
`Instruction` sounds usless here and below (you wouldn't name `add(a, b)` `addFunction(a, b)` ?)

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,460 @@
+ * value will be concatenated into the final translation.
+ */
+export type I18nExpInstruction = number | string;
+/** Mapping of placeholder names to their absolute index in the current view. */
```
is this true for expression / embedded templates ?

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,460 @@
+
+/**
+ * Takes a translation string, the initial list of placeholders (elements and expressions) and the
+ * indexes of their corresponding expression nodes to return a list of instructions for each
```
`indexes of their corresponding expression nodes` unclear, as already commented

vicb replied:
see the design spec.

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,460 @@
+/**
+ * Takes a translation string, the initial list of placeholders (elements and expressions) and the
+ * indexes of their corresponding expression nodes to return a list of instructions for each
+ * template.
```
template ->template function ?

vicb commented on packages/core/src/render3/i18n.ts:
```diff
@@ -0,0 +1,332 @@
+ * Takes a translation string and the initial list of expressions and returns a list of instructions
+ * that will be used to translate an attribute.
+ * Even indexes contain static strings, while odd indexes contain the index of the expression whose
+ * value will be concatenated into the final translation.
```
+1 - I had the same comment before

vicb commented on packages/core/test/render3/compiler_canonical/i18n_spec.ts:
```diff
@@ -0,0 +1,93 @@
+    }
+  });
+
+  it('should support expressions', () => {
```
add html to assert html vs expression indexes

ocombe replied:
there is already a test for that: "should support both html elements, expressions and expressions in attributes"

vicb commented on packages/core/test/render3/compiler_canonical/i18n_spec.ts:
```diff
@@ -0,0 +1,93 @@
+    }
+  });
+
+  it('should support expressions in attributes', () => {
```
add a test for attributes without expression (unless there is one already ?)

vicb replied:
missing a test for embedded templates

ocombe replied:
Attributes without expressions are static attributes, it's already tested in the compiler (nothing required at runtime).
There is a test for embedded templates line ~441

vicb commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -0,0 +1,1190 @@
+    // Html tags are replaced by placeholders.
+    // Open tag placeholders are never re-used (closing tag placeholders can be).
+    const MSG_DIV_SECTION_1 = `{$p_3}trad 1{$c_p}{$p_0}trad 2{$p_1}trad 3{$c_p}{$c_p}`;
+    const i18n_1 = i18nMapping(MSG_DIV_SECTION_1, [{p_0: 1, p_1: 2, p_2: 3, p_3: 4}]);
```
change the placeholder name: if `<tag> / </tag>` -> `START_TAG` and `END_TAG`

vicb commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -0,0 +1,1190 @@
+            '<div><a>trad expr 1</a>hello<b title="start expr 2 middle expr 1 end"><c>trad</c></b></div>');
+  });
+
+  describe('containers', () => {
```
view containers / embedded templates

vicb commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -0,0 +1,1190 @@
+
+    it('should support sibling embedded templates', () => {
+      const MSG_DIV_SECTION_1 = `{$p_0}valeur: {$exp_1}!{$c_p}{$p_1}valeur bis: {$exp_2}!{$c_p}`;
+      // The indexes are based on each template function
```
There should be a canonical to assert that the compiler generates the correct code

vicb commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -0,0 +1,1190 @@
+          // Initial template:
+          // <ul i18n>
+          //   <li *ngFor="let item of items">value: {{item}}</li>
+          // </ul>
```
for each test, please give:
- the initial structure (the lines above),
- the translation.

mhevery commented:
http://test/OCL:201522516:BASE:201522609:1529591995239:2f0e49c4

ocombe committed:
fixup! feat(ivy): runtime i18n

ocombe commented:
new fixup commit:
- renamed all tags in the tests to `START_...` / `END_...`
- added a canonical test for embedded templates

mary-poppins commented:
You can preview fe62b7b at https://pr24037-fe62b7b.ngbuilds.io/.

mary-poppins commented:
You can preview fad434f at https://pr24037-fad434f.ngbuilds.io/.
